File: android/app/src/main/java/com/guardforce/guard/domain/Duty.kt
package com.guardforce.guard.domain

import com.guardforce.guard.api.ShiftInfo
import java.time.Duration
import java.time.Instant
import java.time.OffsetDateTime

/** What the guard sees on the home card. Derived from server state plus what is queued locally. */
sealed class DutyState {
    /** Nothing rostered; ad-hoc start allowed when [canStartAdHoc]. */
    data class NoShift(val canStartAdHoc: Boolean) : DutyState()
    /** A rostered shift more than [startWindow] ahead. */
    data class Upcoming(val shift: ShiftInfo) : DutyState()
    /** Rostered shift inside the start window. */
    data class ReadyToStart(val shift: ShiftInfo) : DutyState()
    /** Check-in saved on the phone, not yet confirmed by the server. */
    data class StartingOffline(val shift: ShiftInfo?, val siteId: String, val sinceMs: Long) : DutyState()
    data class OnDuty(val shift: ShiftInfo?, val serverShiftId: String?, val siteId: String, val sinceMs: Long) : DutyState()
    /** Check-out saved on the phone, waiting for the server. */
    data class EndingOffline(val shift: ShiftInfo?) : DutyState()
    data class Ended(val shift: ShiftInfo) : DutyState()
}

/** Minimal view of the locally tracked shift (see data/db/LocalShiftEntity). */
data class LocalShiftView(val key: String, val serverId: String?, val siteId: String, val startedAtMs: Long, val status: String) {
    companion object {
        const val PENDING_CHECK_IN = "pending_check_in"
        const val ACTIVE = "active"
        const val PENDING_CHECK_OUT = "pending_check_out"
        const val CLOSED = "closed"
        const val FAILED = "failed"
    }
}

object Duty {
    /** A shift may be started this long before its scheduled start. */
    val startWindow: Duration = Duration.ofHours(2)
    /** …and this long after (the server accepts ±6 h; the UI is stricter to avoid mistakes). */
    val lateWindow: Duration = Duration.ofHours(6)

    fun parse(ts: String?): Instant? = ts?.let { runCatching { OffsetDateTime.parse(it).toInstant() }.getOrNull() }

    fun state(shifts: List<ShiftInfo>, activeServerId: String?, local: LocalShiftView?, homeSiteId: String?, now: Instant): DutyState {
        val byId = shifts.associateBy { it.id }
        if (local != null && local.status != LocalShiftView.CLOSED && local.status != LocalShiftView.FAILED) {
            val shift = local.serverId?.let { byId[it] }
            return when (local.status) {
                LocalShiftView.PENDING_CHECK_IN -> DutyState.StartingOffline(shift, local.siteId, local.startedAtMs)
                LocalShiftView.PENDING_CHECK_OUT -> DutyState.EndingOffline(shift)
                else -> DutyState.OnDuty(shift, local.serverId, local.siteId, local.startedAtMs)
            }
        }
        val active = activeServerId?.let { byId[it] } ?: shifts.firstOrNull { it.status == "in_progress" }
        if (active != null) {
            val since = parse(active.start_captured_at ?: active.started_at)?.toEpochMilli() ?: now.toEpochMilli()
            return DutyState.OnDuty(active, active.id, active.site_id, since)
        }
        val scheduled = shifts.filter { it.status == "scheduled" }
            .mapNotNull { s -> parse(s.scheduled_start)?.let { it to s } }
            .sortedBy { (start, _) -> Duration.between(now, start).abs() }
        val next = scheduled.firstOrNull { (start, _) ->
            !start.minus(startWindow).isAfter(now) && !start.plus(lateWindow).isBefore(now)
        }
        if (next != null) return DutyState.ReadyToStart(next.second)
        val upcoming = scheduled.firstOrNull { (start, _) -> start.isAfter(now) }
        if (upcoming != null) return DutyState.Upcoming(upcoming.second)
        val ended = shifts.filter { it.status in setOf("completed", "void_location_off") }
            .maxByOrNull { parse(it.ended_at)?.toEpochMilli() ?: 0L }
        if (ended != null && parse(ended.ended_at)?.let { Duration.between(it, now) < Duration.ofHours(12) } == true) {
            return DutyState.Ended(ended)
        }
        return DutyState.NoShift(canStartAdHoc = homeSiteId != null)
    }
}

File: guard-app/src/domain/duty.ts
import type { ShiftInfo } from "../api/types";

export type LocalShift = { key: string; serverId: string | null; siteId: string; startedAtMs: number; status: LocalStatus };
export type LocalStatus = "pending_check_in" | "active" | "pending_check_out" | "closed" | "failed";

export type DutyState =
  | { kind: "no_shift"; canStartAdHoc: boolean }
  | { kind: "upcoming"; shift: ShiftInfo }
  | { kind: "ready"; shift: ShiftInfo }
  | { kind: "starting_offline"; shift: ShiftInfo | null; siteId: string; sinceMs: number }
  | { kind: "on_duty"; shift: ShiftInfo | null; serverShiftId: string | null; siteId: string; sinceMs: number }
  | { kind: "ending_offline"; shift: ShiftInfo | null }
  | { kind: "ended"; shift: ShiftInfo };

const H = 3_600_000;
export const startWindowMs = 2 * H; // may start this long before scheduled_start
export const lateWindowMs = 6 * H; // …and this long after
const parse = (s?: string | null) => (s ? Date.parse(s) : NaN);

/** What the home card shows: server state overlaid with what this phone has queued. */
export function dutyState(shifts: ShiftInfo[], activeServerId: string | null | undefined, local: LocalShift | null, homeSiteId: string | null | undefined, nowMs: number): DutyState {
  const byId = new Map(shifts.map((s) => [s.id, s]));
  if (local && local.status !== "closed" && local.status !== "failed") {
    const shift = local.serverId ? byId.get(local.serverId) ?? null : null;
    if (local.status === "pending_check_in") return { kind: "starting_offline", shift, siteId: local.siteId, sinceMs: local.startedAtMs };
    if (local.status === "pending_check_out") return { kind: "ending_offline", shift };
    return { kind: "on_duty", shift, serverShiftId: local.serverId, siteId: local.siteId, sinceMs: local.startedAtMs };
  }
  const active = (activeServerId && byId.get(activeServerId)) || shifts.find((s) => s.status === "in_progress");
  if (active) {
    const since = parse(active.start_captured_at ?? active.started_at);
    return { kind: "on_duty", shift: active, serverShiftId: active.id, siteId: active.site_id, sinceMs: Number.isNaN(since) ? nowMs : since };
  }
  const scheduled = shifts.filter((s) => s.status === "scheduled" && !Number.isNaN(parse(s.scheduled_start)))
    .sort((a, b) => Math.abs(parse(a.scheduled_start) - nowMs) - Math.abs(parse(b.scheduled_start) - nowMs));
  const ready = scheduled.find((s) => { const st = parse(s.scheduled_start); return st - startWindowMs <= nowMs && st + lateWindowMs >= nowMs; });
  if (ready) return { kind: "ready", shift: ready };
  const upcoming = scheduled.find((s) => parse(s.scheduled_start) > nowMs);
  if (upcoming) return { kind: "upcoming", shift: upcoming };
  const ended = shifts.filter((s) => s.status === "completed" || s.status === "void_location_off")
    .sort((a, b) => (parse(b.ended_at) || 0) - (parse(a.ended_at) || 0))[0];
  if (ended && nowMs - (parse(ended.ended_at) || 0) < 12 * H) return { kind: "ended", shift: ended };
  return { kind: "no_shift", canStartAdHoc: !!homeSiteId };
}
